torchdata/datapipes/iter/util/randomsplitter.py

```diff
@@ -0,0 +1,171 @@
+            return [SplitterIterator(container, k) for k in list(weights.keys())]
+        else:
+            if target in weights.keys():
+                return SplitterIterator(container, target)
+            else:
+                raise KeyError(f"`target={target}` does not match any key in `weights`.")
+
+
+class _RandomSplitterIterDataPipe(IterDataPipe):
+    def __init__(
+        self,
+        source_datapipe: IterDataPipe,
+        total_length: int,
+        weights: Dict[T, Union[int, float]],
+        seed,
+    ):
+        self.source_datapipe: IterDataPipe = source_datapipe
+        self.total_length = total_length
+        self._seed = seed
+        self.norm_weights = self.normalize_weights(weights, total_length)
+        self.keys = list(weights.keys())
+        self.key_to_index = {k: i for i, k in enumerate(self.keys)}
+        self.weights = [self.norm_weights[k] for k in self.keys]
+        self._rng = random.Random(self._seed)
+
+    def draw(self):
+        selected_key = self._rng.choices(self.keys, self.weights)[0]
+        self.weights[self.key_to_index[selected_key]] -= 1
+        return selected_key
+
+    @staticmethod
+    def normalize_weights(weights, total_length: int):
+        total_weight = sum(weights.values())
+        return {k: round(float(w) * total_length / total_weight) for k, w in weights.items()}
+
+    def reset(self):
+        self._rng = random.Random(self._seed)
+        self.weights = [self.norm_weights[k] for k in self.keys]
+
+    def override_seed(self, seed):
+        """
+        Update the `seed`. The new `seed` will be used in the next iteration.
+        """
+        self._seed = seed
+        return self
+
+    def __getstate__(self):
+        if IterDataPipe.getstate_hook is not None:
+            return IterDataPipe.getstate_hook(self)
+        state = (
+            self.source_datapipe,
+            self.total_length,
+            self._seed,
+            self.norm_weights,
+            self.keys,
+            self.key_to_index,
+            self.weights,
+            self._rng.getstate(),
+        )
+        return state
+
+    def __setstate__(self, state):
+        (
+            self.source_datapipe,
+            self.total_length,
+            self._seed,
+            self.norm_weights,
+            self.keys,
+            self.key_to_index,
+            self.weights,
+            rng_state,
+        ) = state
+        self._rng = random.Random()
+        self._rng.setstate(rng_state)
+
+
+class SplitterIterator(IterDataPipe):
+    def __init__(self, main_datapipe: _RandomSplitterIterDataPipe, target: T):
+        self.main_datapipe = main_datapipe
+        self.target = target
+
+    def __iter__(self):
+        self.main_datapipe.reset()
+        for sample in self.main_datapipe.source_datapipe:
+            if self.main_datapipe.draw() == self.target:
+                yield sample
+
+    def __len__(self):
+        return self.main_datapipe.norm_weights[self.target]
```

Comment:
nit: Would it be better to add a function call to get length rather than hard code the attribute name?

Reply:
Actually I realized the length cannot be known in advance, so I'm taking the method out entirely.

Example: 50/50 split for a DP of length 7. You don't know if `dp1` or `dp2` will have length 3 or 4.

Reply:
Makes sense.